tox/venv.py

```diff
@@ -416,6 +418,25 @@ def getdigest(path):
     return path.computehash()
 
 
+def shebang_wrapper(args):
+    # handle interpreter directive with long lengths
+    #
+    # When preparing virtual environments in a file container which has large
+    # length, the system might not be able to invoke shebang scripts which
+    # define interpreters beyond system limits (i.e. Linux as a limit of 128;
+    # BINPRM_BUF_SIZE). This wrapper can be used to check if the executable is
+    # a script containing a shebang line. If so, read the first line for the
+    # interpreter and check if length exceeds "common" limit. If the interpreter
+    # exceeds the limit, adjust the argument list to directly call the
+    # interpreter.
+    with open(args[0], 'rb') as f:
+        if f.read(1) == '#' and f.read(1) == '!':
+            interp = f.readline(1024).rstrip()
+            if len(interp) > 125:
```

Comment:
Why is it not possible to take the actual value of `BINPRM_BUF_SIZE` into account here?

Reply:
Truthfully, it was a guess to what the actual total interpret length could be 😅. I suspected that `BINPRM_BUF_SIZE`'s value (i.e. 128) couldn't be used directly since I didn't know (or explicitly check) to whether or not the prefix `#!` and other conditions would be part of the "real limit". But yes, I would agree just throwing it this magic number here isn't really helpful for maintenance reasons. 

I re-checked Linux's [`binfmt_script.c`](https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux-stable.git/tree/fs/binfmt_script.c?h=linux-4.16.y#n24) and (I believe) confirmed that a value of `125` should be the value. The define `BINPRM_BUF_SIZE` (128) with an enforced null-terminating character gives us `127` characters. The prefix `#!` does contribute to the maximum length, so we have a maximum of `125` characters for our interpreter string (at least, on Linux; [1]). Instead of just having a value `127`, we could do something like (a note why `MAXINTERP`'s value exist in the comment):

```
MAXINTERP = 125
if len(interp) > MAXINTERP:
```
---

On a related note, it may be ideal to also change `args = [interp] + args` to be `args = [interp.lstrip()] + args`. This should trim leading whitespace characters from the interpreter before execution (which I assume is a good thing).

[1]: https://www.in-ulm.de/~mascheck/various/shebang/#length

Reply:
The problem is that AFAIK different distros patch this length differently, so we would need to be able to know from within a running system. Another (ugly) approach would be to try it and react to that specific problem to try again with the adjusted approach ...

Reply:
Agreed that handling interpreter limits for various distributions is a problem. The lazy'er approach would be to just define the common maximum interpret length and allow the user to override the limit (through an environment variable?) to help force this workaround in a more limited-length environment.

The "ugly" approach could work. Execution the command as expected and if it fails with a `ENOENT` (and is !win32 and is a detected shebang script), try it again with the wrapped call (I'd worry if an executable actually returns a value `ENOENT` for some operating case; but what are the chances there...).

Another approach would be just to detect is the to be executed script is a shebang script and always wrap it.

---

The code `interp = f.readline(1024).rstrip()` probably needs more work as well:
 - In the event that we don't care about the interpreter's length, we could drop `rstrip`.
 - The size `1024` may be too small. In FreeBSD, `MAXINTERP` is set to `PATH_MAX` (2048).
 - The directive should probably be split on whitespaces (to handle scenarios where something like `#!/usr/bin/env python` set).